### PR TITLE
instructor remind students: allow even if the session is closed #7073

### LINKED ISSUE #7073: instructor remind students: allow even if the session is closed
Current situation:
The `remind` button in instructor home page is disabled if the session is closed.
image

Suggestion: 
Enable it even if the session is closed, but not if it is awaiting.

Justification: 
In rare cases, a student might want to have a copy of the link even after the session is closed e.g. to check his/her submission

### PR DESCRIPTION
**Pull request by sujeet14108**
Fixes #7073 

screenshot from 2017-04-09 19-34-24

### CODE REVIEW

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 1]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
...
-                                                                 feedbackSessionClosed.getFeedbackSessionName()));
+
+        ______TS("remind action: CLOSED feedback session - inner button");
+
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickAndCancel(homePage.getRemindInnerLink(feedbackSessionClosed.getCourseId(),
+                                                            feedbackSessionOpen.getFeedbackSessionName()));
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickAndConfirm(homePage.getRemindInnerLink(feedbackSessionClosed.getCourseId(),
+                                                             feedbackSessionClosed.getFeedbackSessionName()));
```
Comment (AnaghHegde) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Why these are repeated ?. Also now we are using 8 space indentation. You should break at higher level. 
After this `homePage.clickAndConfirm(homePage.getRemindInnerLink(`

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 2]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
...
+        homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_REMINDERSSENT);
+
+        //go back to previous page because 'send reminder' redirects to the 'Feedbacks' page.
+        homePage.goToPreviousPage(InstructorHomePage.class);
+
+        ______TS("remind particular users action: CLOSED feedback session");
+
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickRemindParticularUsersLink(feedbackSessionClosed.getCourseId(),
+                                                feedbackSessionClosed.getFeedbackSessionName());
```
Comment (AnaghHegde) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: same as above use 8 space indentation

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 3]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
...
+        homePage.waitForAjaxLoaderGifToDisappear();
+        homePage.submitRemindParticularUsersForm();
+        homePage.waitForPageToLoad();
+        homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_REMINDERSEMPTYRECIPIENT);
+        homePage.goToPreviousPage(InstructorHomePage.class);
+
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickRemindParticularUsersLink(feedbackSessionClosed.getCourseId(),
+                                                feedbackSessionClosed.getFeedbackSessionName());
```
Comment (AnaghHegde) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: same as above.

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java (hunk 1 above)
Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: @AnaghHegde  It is here to check if remind button is  working when session is closed as done in above section for open session

Comment (AnaghHegde) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Take a look at line 280,281 and 283,284 those are same. I am asking why they are repeated in line 283 and 284.

Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Line 281 and 284 are not the same. One is `clickAndCancel`, another is `clickAndConfirm`.

**Comment by whipermr5** [comment]
@sujeet14108 You need to specify UTC timezone as a VM argument for GodMode to write the correct times. Refer [here](https://github.com/TEAMMATES/teammates/blob/master/docs/development.md#testing).

**Comment by sujeet14108** [comment]
@whipermr5 
I am facing a problem in testing remind button for reminding particular users. In the close session no students are there so unable to test that feature.
screenshot from 2017-04-12 03-29-37

But for open session of same course there are students so testing is fine.
So how can i add students to that session for testing.

`        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
        homePage.clickRemindParticularUsersLink(feedbackSessionClosed.getCourseId(),
                                                feedbackSessionClosed.getFeedbackSessionName());
        homePage.waitForAjaxLoaderGifToDisappear();
        homePage.fillRemindParticularUsersForm();
        homePage.submitRemindParticularUsersForm();
        homePage.waitForPageToLoad();
        homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_REMINDERSSENT);
        homePage.goToPreviousPage(InstructorHomePage.class);`

**Comment by whipermr5** [comment]
@sujeet14108 Looks like students are based on the target respondents of questions in the feedback session - e.g. if you add a question where students give feedback, students will appear in the remind list; if the question involves instructors giving feedback, instructors will appear. You'll have to add a feedback question where students give feedback to the test data for the closed feedback session(s). The relevant file is `InstructorHomePageUiTest2.json`.

**Comment by sujeet14108** [comment]
@whipermr5 
For **Published session** do we want remind button enable or disabled?

**Comment by sujeet14108** [comment]
@whipermr5  Ready for review

**Comment by whipermr5** [comment]
@damithc Just to confirm - the remind button should not be enabled if the session is already published right? @sujeet14108's decision makes sense as students would already have received the email when the results were published.

**Comment by damithc** [comment]
> @damithc Just to confirm - the remind button should not be enabled if the session is already published right? @sujeet14108's decision makes sense as students would already have received the email when the results were published.

Resending the link after a session is published is still a valid use case (the student may not have received any of the previous emails). However, we can tackle that separately because the email should contain both submission and result links.

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 4]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
         homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_REMINDERSSENT);
         homePage.goToPreviousPage(InstructorHomePage.class);
 
-        ______TS("remind action: CLOSED feedback session");
 
-        homePage.verifyUnclickable(homePage.getRemindLink(feedbackSessionClosed.getCourseId(),
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Can we use this example to `verifyUnclickable` for the remind links of awaiting and published sessions?

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: @whipermr5  For Published its already done below closed seesion

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: And awaiting on the top

Comment (whipermr5) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: My bad; I didn't realise it's already in the test 😅

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 5]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
...
-                                                          feedbackSessionClosed.getFeedbackSessionName()));
-        homePage.verifyUnclickable(homePage.getRemindOptionsLink(feedbackSessionClosed.getCourseId(),
-                                                                 feedbackSessionClosed.getFeedbackSessionName()));
+
+        ______TS("remind action: CLOSED feedback session - inner button");
+
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickAndCancel(homePage.getRemindInnerLink(feedbackSessionClosed.getCourseId(),
+                                                            feedbackSessionOpen.getFeedbackSessionName()));
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: suggested line wrapping (use 8 space indentation):
```java
homePage.clickAndCancel(homePage.getRemindInnerLink(
        feedbackSessionClosed.getCourseId(), feedbackSessionOpen.getFeedbackSessionName()));
```

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: @damithc  should i correct the indentation in other methods which i have not edited also like in 

             ______TS("remind action: PUBLISHED feedback session");

        homePage.verifyUnclickable(homePage.getRemindLink(feedbackSessionPublished.getCourseId(),
                                                          feedbackSessionPublished.getFeedbackSessionName()));
        homePage.verifyUnclickable(homePage.getRemindOptionsLink(feedbackSessionPublished.getCourseId(),
                                                                 feedbackSessionPublished.getFeedbackSessionName()));`

Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: No, just the lines you touched.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Done

**Comment by sujeet14108** [comment]
@damithc  Done :+1:

File: src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java [hunk 6]
```diff
@@ -273,12 +273,47 @@ private void testRemindActions() {
...
+        homePage.clickAndCancel(homePage.getRemindInnerLink(feedbackSessionClosed.getCourseId(),
+                feedbackSessionOpen.getFeedbackSessionName()));
+        homePage.clickRemindOptionsLink(feedbackSessionClosed.getCourseId(), feedbackSessionClosed.getFeedbackSessionName());
+        homePage.clickAndConfirm(homePage.getRemindInnerLink(feedbackSessionClosed.getCourseId(),
+                feedbackSessionClosed.getFeedbackSessionName()));
+        homePage.waitForPageToLoad();
+        homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_REMINDERSSENT);
+
+        //go back to previous page because 'send reminder' redirects to the 'Feedbacks' page.
+        homePage.goToPreviousPage(InstructorHomePage.class);
```
Comment (damithc) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: Is this needed? I think we changed this behavior some time ago. Now the browser remains in the home page after sending a reminder, does it not? If it is no longer needed, remove it from the tests you added and create another issue to remove it from other similar cases in this class.

Comment (sujeet14108) [inline src/test/java/teammates/test/cases/browsertests/InstructorHomePageUiTest.java]: @damithc Yes this is not needed. I will correct it for my test and create  issue for other test.

**Comment by sujeet14108** [comment]
@damithc Done :+1:  Please Check
I am crating issue for it and will solve it.

**Pull request merged**